### PR TITLE
client-go: fix fixture data race

### LINKED ISSUE #103148: client-go: fixture data race
#### What happened:
Hello,
Found a potential data race in `client-go` testing.
There is a test `TestWatchCallNonNamespace`
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/testing/fixture_test.go#L56
If we update it like this (we actually can do it [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/testing/fixture.go#L97)):
```
...
go func() {
	err := o.Create(testResource, testObj, ns)
	if err != nil {
		t.Errorf("test resource creation failed: %v", err)
	}
	// Potential data race
	_, err = o.Get(testResource, ns, testObj.GetName())
	if err != nil {
		t.Errorf("test resource getting failed: %v", err)
	}
}()
out := <-watch.ResultChan()
// Modify received object
// Potential data race
outCasted := out.Object.(*unstructured.Unstructured)
outCasted.Object["data"] = strconv.Itoa(rand.Int())
...
```
If watcher updates received object we get a data race.

#### What you expected to happen:
No data race

#### Anything else we need to know?:
My suggestion: send copy of object here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/testing/fixture.go#L403
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/testing/fixture.go#L419

### PR DESCRIPTION
**Pull request by novahe**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
watcher updates received object we get a data race.
#### Which issue(s) this PR fixes:

Fixes #103148

#### Special notes for your reviewer:
There may be a better way than copying object
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by denis-tingaikin — APPROVED** [review]

File: staging/src/k8s.io/client-go/testing/fixture.go [hunk 1]
```diff
@@ -416,7 +417,8 @@ func (t *tracker) add(gvr schema.GroupVersionResource, obj runtime.Object, ns st
 	t.objects[gvr][namespacedName] = obj
 
 	for _, w := range t.getWatches(gvr, ns) {
-		w.Add(obj)
+		// To avoid the object from being accidentally modified by watcher
+		w.Add(obj.DeepCopyObject())
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/testing/fixture.go:421]: probably also needed in Delete when calling w.Delete(obj)

Comment (novahe) [reply staging/src/k8s.io/client-go/testing/fixture.go:421]: We store obj in `t.objects` and do `w.Delete(obj)`, obj is deleted from `t.objects` map

https://github.com/kubernetes/kubernetes/blob/3f86955da2e55ad6e7fe025cdc705b94e03a0357/staging/src/k8s.io/client-go/testing/fixture.go#L453-L461

watcher receives obj that is not stored in `t.objects` map.

and I also test following code with `go test -race .`：
```go
	go func() {
		err := o.Create(testResource, testObj, ns)
		if err != nil {
			t.Errorf("test resource creation failed: %v", err)
		}
		_, err = o.Get(testResource, ns, testObj.GetName())
		// Delete obj
		err = o.Delete(testResource, ns, testObj.GetName())
		if err != nil {
			t.Errorf("test resource getting failed: %v", err)
		}

	}()
	out := <-watch.ResultChan()
	// Modify received object
	// Potential data race
	outCasted := out.Object.(*unstructured.Unstructured)
	outCasted.Object["data"] = strconv.Itoa(rand.Int())
```
So I think there is no data race :), What did I miss?

Comment (liggitt) [reply staging/src/k8s.io/client-go/testing/fixture.go:421]: if there are multiple watchers, and they modify the object in their delete handler, they should not observe each other's modifications

Comment (novahe) [reply staging/src/k8s.io/client-go/testing/fixture.go:421]: Thank you, i get it and fix it. PTAL

**Pull request merged**